RCHAIN-3220: Rholang interpreter prints out unmatched sends

tymm opened the pull request:
## Overview
This is part 1.
Part 2 will add support for RhoSpec.

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3672


### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>



### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

tymm committed:
Fix storage output for grpc clients

tymm committed:
Add StoragePrinter.prettyPrintUnmatchedSends

tymm committed:
Add possibility of printing unmatched sends for `rnode eval`

tymm committed:
Fix RholangCLI by adding merge strategy

tymm committed:
Add possibility of printing unmatched sends for RholangCLI

kayvank commented on node/src/main/scala/coop/rchain/node/api/ReplGrpcService.scala:
```diff
@@ -67,7 +70,7 @@ private[api] class ReplGrpcService(runtime: Runtime[Task], worker: Scheduler)
             s"Deployment cost: $cost\n" +
-              s"${errorStr}Storage Contents:\n ${StoragePrinter.prettyPrint(runtime.space)}"
+              s"${errorStr}Storage Contents:\n$prettyStorage"
 
```
why an empty line?

tymm replied:
Not sure what you mean?
Before the first term was offset by one space while all following terms were not.

kayvank replied:
just a formatting question. Why keep the empty lines, line 74 & line 75?

tymm replied:
Oh, I see. Just removed them.

kayvank commented on node/src/main/scala/coop/rchain/node/effects/ReplClient.scala:
```diff
@@ -43,15 +46,18 @@ class GrpcReplClient(host: String, port: Int, maxMessageSize: Int)
+      .traverse(eval(_, unmatchedSends))
 
-  def eval(fileName: String): Task[Either[Throwable, String]] = {
+  def eval(fileName: String, printUnmatchedSendsOnly: Boolean): Task[Either[Throwable, String]] = {
```
should this private?

tymm replied:
Nope, it's called from `ReplClient.scala`.

tymm committed:
Make `RuntimeManager.withRuntimeLock` public

tymm committed:
Add possibility of printing unmatched sends for `rnode propose`

When used this will print the unmatched sends per deploy in the propose. Deploys won't influence the output of other deploys in the proposed block (unlike to what happens on-chain).

tymm committed:
Add tests for StoragePrinter.prettyPrintUnmatchedSends

odeac commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/StoragePrinter.scala:
```diff
@@ -50,6 +50,30 @@ object StoragePrinter {
+            case Row(data: Seq[Datum[ListParWithRandom]], _) =>
+              toSends(data)(channels)
+          }
+      }.toList
```
do you need the toList here since the next step is to fold it?

odeac commented on node/src/main/scala/coop/rchain/node/api/ReplGrpcService.scala:
```diff
@@ -57,7 +57,9 @@ private[api] class ReplGrpcService(runtime: Runtime[Task], worker: Scheduler)
               Interpreter[Task].evaluate(runtime, source, NormalizerEnv.Empty)
             }
-            prettyStorage                <- StoragePrinter.prettyPrint(runtime.space)
+            prettyStorage <- if (printUnmatchedSendsOnly)
```
So we can choose to either print only the tuplespace or only the unmatched sends. Does it makes sense to print both in some situations?

tymm replied:
When you print the tuplespace, you also get the unmatched sends. Therefore I would argue that it doesn't make sense to print both at the same time.

odeac replied:
ok

odeac commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/RholangCLI.scala:
```diff
@@ -226,7 +252,12 @@ object RholangCLI {
+  def evaluatePar(
+      runtime: Runtime[Task],
+      source: String,
+      quiet: Boolean = false,
```
are the default values  required for the parameters quiet and unmatchedSendsOnly? My impression is that, at call site, we provide values for both

odeac commented on casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala:
```diff
@@ -566,20 +572,44 @@ object BlockAPI {
+      casper: MultiParentCasper[F],
+      deploys: Seq[DeployData]
+  ): F[Option[String]] =
+    casper.getRuntimeManager >>= {
```
you should be able to use traverse instead of flatMap + match

tymm replied:
Hmm, how would that work? We would have to add the `Traverse` typeclass to `F`. Do we want that?

odeac replied:
I'm sorry. I wasn't paying attention at the types when I wrote my comment.

This is what I meant:

```scala
  private def prettyPrintUnmatchedSends[F[_]: Concurrent](
      casper: MultiParentCasper[F],
      deploys: Seq[DeployData]
  ): F[Option[String]] =
    casper.getRuntimeManager >>= (maybeRuntime =>
      maybeRuntime.traverse( rm =>
        rm.withRuntimeLock(
          runtime => StoragePrinter.prettyPrintUnmatchedSends(deploys, runtime))))
```

odeac commented on casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala:
```diff
@@ -566,20 +572,44 @@ object BlockAPI {
+  ): F[Option[String]] =
+    casper.getRuntimeManager >>= {
+      case Some(rm) =>
+        rm.withRuntimeLock(runtime => StoragePrinter.prettyPrintUnmatchedSends(deploys, runtime))
```
now I see the usage of withRuntimeLock but why do you need to lock the runtime? I'm thinking that the printing of unmatched sends happen after the execution ended so it's not executed concurrently with other runtime operations

tymm replied:
`StoragePrinter.prettyPrintUnmatchedSends` executes Rholang. Therefore I think we should use `withRuntimeLock`.

odeac replied:
ok

odeac commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/StoragePrinter.scala:
```diff
@@ -40,39 +45,79 @@ object StoragePrinter {
+          prettyPrintUnmatchedSends(deploy, runtime)
+            .map(unmatchedSends => (deploy, unmatchedSends))
+      )
+      .toList
```
I assume if you need to traverse the collection so you can use toStream instead of toList. That would be better because it doesn't build the intermediary list

odeac commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/StoragePrinter.scala:
```diff
@@ -40,39 +45,79 @@ object StoragePrinter {
+            .map(unmatchedSends => (deploy, unmatchedSends))
+      )
+      .toList
+      .sequence
```
map followed by sequence can be replaced by traverse

tymm commented:
bors r+

tymm commented:
bors r+

tymm commented:
bors try

tymm commented:
bors retry

tymm commented:
bors r+

tymm commented:
bors r+

tymm commented:
bors try

tymm commented:
bors r+
